File: android-core/src/androidTest/java/com/mparticle/networking/MParticleBaseClientImplTest.java
package com.mparticle.networking;

import com.mparticle.MParticle;
import com.mparticle.MParticleOptions;
import com.mparticle.internal.AccessUtils;
import com.mparticle.testutils.BaseCleanInstallEachTest;

import org.junit.Before;
import org.junit.Test;

import java.net.MalformedURLException;
import java.util.HashMap;
import java.util.Map;

import static junit.framework.TestCase.assertTrue;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;

public class MParticleBaseClientImplTest extends BaseCleanInstallEachTest {
    Map<MParticleBaseClientImpl.Endpoint, MPUrl> defaultUrls = new HashMap<MParticleBaseClientImpl.Endpoint, MPUrl>();
    String apiKey = mRandomUtils.getAlphaString(10);

    @Before
    public void before() throws InterruptedException, MalformedURLException {
        startMParticle(MParticleOptions.builder(mContext).credentials(apiKey, "secret"));
        MParticleBaseClientImpl baseClientImpl = (MParticleBaseClientImpl)AccessUtils.getApiClient();
        for (MParticleBaseClientImpl.Endpoint endpoint: MParticleBaseClientImpl.Endpoint.values()) {
            defaultUrls.put(endpoint, baseClientImpl.getUrl(endpoint, endpoint.name()));
        }
        MParticle.setInstance(null);
    }

    @Test
    public void testGetUrlForceDefaultOption() throws MalformedURLException {
        String identityUrl = mRandomUtils.getAlphaString(20);
        String configUrl = mRandomUtils.getAlphaString(20);
        String audienceUrl = mRandomUtils.getAlphaString(20);
        String eventsUrl = mRandomUtils.getAlphaString(20);

        MParticleOptions options = MParticleOptions.builder(mContext)
                .credentials(apiKey, "secret")
                .networkOptions(NetworkOptions.builder()
                        .addDomainMapping(DomainMapping.audienceMapping(audienceUrl)
                                .build())
                        .addDomainMapping(DomainMapping.configMapping(configUrl)
                                .build())
                        .addDomainMapping(DomainMapping.identityMapping(identityUrl)
                                .build())
                        .addDomainMapping(DomainMapping.eventsMapping(eventsUrl).build())
                        .build())
                .build();
        MParticle.start(options);

        MParticleBaseClientImpl baseClientImpl = (MParticleBaseClientImpl)AccessUtils.getApiClient();

        for (MParticleBaseClientImpl.Endpoint endpoint: MParticleBaseClientImpl.Endpoint.values()) {
            MPUrl generatedUrl = baseClientImpl.getUrl(endpoint, endpoint.name(), true);
            assertEquals(defaultUrls.get(endpoint).toString(), generatedUrl.toString());
            assertTrue(generatedUrl == generatedUrl.getDefaultUrl());
        }

        for (MParticleBaseClientImpl.Endpoint endpoint: MParticleBaseClientImpl.Endpoint.values()) {
            MPUrl generatedUrl = baseClientImpl.getUrl(endpoint, endpoint.name(), false);
            assertNotEquals(defaultUrls.get(endpoint).toString(), generatedUrl.toString());
            assertFalse(generatedUrl == generatedUrl.getDefaultUrl());
            assertEquals(defaultUrls.get(endpoint).toString(), generatedUrl.getDefaultUrl().toString());
        }
    }





}

File: android-core/src/androidTest/java/com/mparticle/networking/MParticleBaseClientImplTest.kt
package com.mparticle.networking

import com.mparticle.MParticle
import com.mparticle.MParticleOptions
import com.mparticle.internal.AccessUtils
import com.mparticle.testing.BaseTest
import com.mparticle.testing.RandomUtils
import com.mparticle.testing.context
import com.mparticle.utils.startMParticle
import junit.framework.TestCase
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.net.MalformedURLException

class MParticleBaseClientImplTest : BaseTest() {
    var defaultUrls: MutableMap<MParticleBaseClientImpl.Endpoint, MPUrl> = HashMap()
    var apiKey: String = "key"
    @Before
    @Throws(InterruptedException::class, MalformedURLException::class)
    fun before() {
        startMParticle(MParticleOptions.builder(context).credentials(apiKey, "secret"))
        val baseClientImpl = AccessUtils.apiClient as MParticleBaseClientImpl
        for (endpoint in MParticleBaseClientImpl.Endpoint.values()) {
            defaultUrls[endpoint] = baseClientImpl.getUrl(endpoint, endpoint.name)
        }
        MParticle.setInstance(null)
    }

    @Test
    @Throws(MalformedURLException::class)
    fun testGetUrlForceDefaultOption() {
        val identityUrl: String = RandomUtils.getAlphaString(20)
        val configUrl: String = RandomUtils.getAlphaString(20)
        val audienceUrl: String = RandomUtils.getAlphaString(20)
        val eventsUrl: String = RandomUtils.getAlphaString(20)
        val options = MParticleOptions.builder(context)
            .credentials(apiKey, "secret")
            .networkOptions(
                NetworkOptions.builder()
                    .addDomainMapping(
                        DomainMapping.audienceMapping(audienceUrl)
                            .build()
                    )
                    .addDomainMapping(
                        DomainMapping.configMapping(configUrl)
                            .build()
                    )
                    .addDomainMapping(
                        DomainMapping.identityMapping(identityUrl)
                            .build()
                    )
                    .addDomainMapping(DomainMapping.eventsMapping(eventsUrl).build())
                    .build()
            )
            .build()
        MParticle.start(options)
        val baseClientImpl = AccessUtils.apiClient as MParticleBaseClientImpl
        for (endpoint in MParticleBaseClientImpl.Endpoint.values()) {
            val generatedUrl = baseClientImpl.getUrl(endpoint, endpoint.name, true)
            Assert.assertEquals(defaultUrls[endpoint].toString(), generatedUrl.toString())
            TestCase.assertTrue(generatedUrl === generatedUrl.defaultUrl)
        }
        for (endpoint in MParticleBaseClientImpl.Endpoint.values()) {
            val generatedUrl = baseClientImpl.getUrl(endpoint, endpoint.name, false)
            Assert.assertNotEquals(defaultUrls[endpoint].toString(), generatedUrl.toString())
            Assert.assertFalse(generatedUrl === generatedUrl.defaultUrl)
            Assert.assertEquals(
                defaultUrls[endpoint].toString(),
                generatedUrl.defaultUrl.toString()
            )
        }
    }
}
